src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlockMeasureCache.MeasureEntry.cs

```diff
@@ -11,66 +11,121 @@ internal partial class TextBlockMeasureCache
 		/// <summary>
 		/// A set of TextBlock measures for a set of text characteristics
 		/// </summary>
-		class MeasureEntry
+		private class MeasureEntry
 		{
-			private Dictionary<CachedSize, MeasureSizeEntry> _sizes
+			private readonly Dictionary<CachedSize, MeasureSizeEntry> _sizes
 				= new Dictionary<CachedSize, MeasureSizeEntry>();
 
-			private LinkedList<CachedSize> _queue = new LinkedList<CachedSize>();
+			private readonly LinkedList<CachedSize> _queue = new LinkedList<CachedSize>();
 
 			public LinkedListNode<MeasureKey> ListNode { get; }
 
 			public MeasureEntry(LinkedListNode<MeasureKey> node) => this.ListNode = node;
 
 			public Size? FindMeasuredSize(MeasureKey key, Size availableSize)
 			{
-				if(_sizes.TryGetValue(CachedTuple.Create(availableSize.Width, availableSize.Height), out var sizeEntry))
+				var currentAvailableWidth = availableSize.Width;
+				var currentAvailableHeight = availableSize.Height;
+
+				if(_sizes.TryGetValue(CachedTuple.Create(currentAvailableWidth, currentAvailableHeight), out var sizeEntry))
 				{
 					return sizeEntry.MeasuredSize;
 				}
-				else
+
+				var isWrapping = key.IsWrapping;
+
+				foreach (var kvp in _sizes)
 				{
-					if(key.TextWrapping == TextWrapping.NoWrap)
+					var size = kvp.Key;
+					var measureSizeEntry = kvp.Value;
+					var measurementCachedWidth = measureSizeEntry.MeasuredSize.Width;
+					var measurementCachedHeight = measureSizeEntry.MeasuredSize.Height;
+					var measurementAvailableWidth = size.Item1;
+					var measurementAvailableHeight = size.Item2;
+
+					if (isWrapping)
+					{
+						if (measurementCachedWidth <= currentAvailableWidth
+							&& currentAvailableWidth <= measurementAvailableWidth)
+						{
+							// Ok we can reuse it
+						}
+						else
+						{
+							continue; // Check for another cached measurement
+						}
+					}
+					else
 					{
-						// No wrap, assume any measured width below the asked available size
-						// is valid, if the available size is greater.
-						foreach(var keySize in _sizes)
+						// Non-wrapping text
+
+						if (double.IsInfinity(measurementAvailableWidth))
+						{
+							// Previous measurement was unconstrained
+							// horizontally: we can definitely reuse it.
+						}
+						else
 						{
-							if(keySize.Key.Item1 >= availableSize.Width &&
-								keySize.Value.MeasuredSize.Width <= availableSize.Width)
+							if (Math.Abs(measurementCachedWidth - measurementAvailableWidth) < 0.5d)
 							{
-								MoveToLast(keySize.Key, keySize.Value);
-								return keySize.Value.MeasuredSize;
+								// This measure was constrained, we can reuse only if the width is the same.
+
+								if (Math.Abs(measurementCachedWidth - currentAvailableWidth) < 0.5d)
```

Comment:
I think it could be really easier to read if you put just a single `if` with the three conditions

Reply:
I separated it precisely to make it clear because it was too difficult to understand.